### ISSUE TITLE
Understanding preds data sent to custom scorer through feval in multi:softprob objective 

### ISSUE THREAD

**Issue by talipini** [dmlc/xgboost#6221]
Data sent to custom feval function looks confusing when training in multi:softprob objective. My expectation is that the data sent is the probabilities of the classes but that doesn't seem to be the case so I am trying to understand what the data is. 

Here is the sample code for discussion purposes. I am just using iris data for this example and introducing a custom feval function to score the preds during the evaluation stage. Is the data sent the probability for each class similar to the probabilities given the fully trained model in prediction stage? Doesn't look like it since I saw few values >1. I think these are leaf values but not sure how to calculate the final probabilities out of it. Or if the max of the values can be considered as the predicted class for that sample. Appreciate any color.

Starting Train
**Preds in custom scorer**
 [[0.281155   0.909091   0.2899053 ]
 [0.9260356  0.28698218 0.2899053 ]
 [0.281155   0.289781   0.92043793]
 [0.281155   0.909091   0.2899053 ]
 [0.281155   0.909091   0.49999988]]
Preds Argmax:[1 0 2 1 1] vs Truths:[1. 0. 2. 1. 1.]
[0]     test-merror:0.00000     test-score:1.00000

**Predicting After Training**
[[0.25756413 0.48260805 0.2598278 ]
 [0.48611447 0.25656724 0.25731832]
 [0.25616166 0.2583809  0.48545748]
 [0.25756413 0.48260805 0.2598278 ]
 [0.24281408 0.45497027 0.3022157 ]]

```
import xgboost
import numpy as np
from sklearn import datasets
from sklearn.model_selection import train_test_split


def custom_scorer(preds, truths): 
    print(f"Preds in scorer \n {preds[:5,]}")
    pred_classes = preds.argmax(axis=-1)
    print(f"Preds Argmax:{pred_classes[:5,]} vs Truths:{truths.get_label()[:5,]} ")
    return "score", np.corrcoef(pred_classes, truths.get_label())[0,1]
 
def run_iris_test(): 
    iris = datasets.load_iris()
    X = iris.data
    y = iris.target

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    dtrain = xgboost.DMatrix(X_train, label=y_train)
    dtest = xgboost.DMatrix(X_test, label=y_test)

    param = {
        'max_depth': 3,  # the maximum depth of each tree
        'eta': 0.3,  # the training step for each iteration
        'objective': 'multi:softprob',  # error evaluation for multiclass training
        'num_class': 3,
        'tree_method':'gpu_hist',  
        }   

    print(f"Starting Train")
    bst = xgboost.train(param, dtrain, 
            num_boost_round=1,
            feval=custom_scorer,
            maximize = True,
            evals=[(dtest, 'test')],
            verbose_eval=1,
            )
    
    print(f"\n\nPredicting After Training ")
    preds = bst.predict(dtest)
    print(preds[:5,])


if __name__ == '__main__':
    run_iris_test()
```

**Comment by talipini** [comment]
I believe the values are sum of leaf values across the trees. To get to the probabilities, we need to apply the softmax on each samples leaf values.  

For ex, applying softmax on the leaf values gets us the probabilities..
[0.281155 0.909091 0.2899053 ]  --> [0.25756 0.48261 0.25983]

use scipy softmax or simply  `np.exp(leaf_values) / np.sum(np.exp(leaf_values), axis=1)`

Leaving it in here, in case it might be useful to some one else.

**Comment by hcho3** [comment]
@talipini Yes, that is correct. The values are the sum of raw scores produced by the leaf nodes in all the trees. You will need to manually apply the softmax.

From the release note of XGBoost 1.2.0:

> **Breaking: Custom evaluation metric now receives raw prediction (#5954)**
> * Previously, the custom evaluation metric received a transformed prediction result when used with a classifier. Now the custom metric will receive a raw (untransformed) prediction and will need to transform the prediction itself. See demo/guide-python/custom_softmax.py for an example.
> * This change is to make the custom metric behave consistently with the custom objective, which already receives raw prediction (#5564).
